api/v1alpha1/stardogrole_types.go

```diff
@@ -25,11 +25,22 @@ import (
 
 // StardogRoleSpec defines the desired state of StardogRole
 type StardogRoleSpec struct {
-	// INSERT ADDITIONAL SPEC FIELDS - desired state of cluster
-	// Important: Run "make" to regenerate code after modifying this file
+    // Name describes the name of a Role.
+    Name string `json:"name,omitempty"`
+    // StardogPermissionSpec describes a list permissions assigned to a role
+    permissions []StardogPermissionSpec `json:"permissions,omitempty"`
+}
 
-	// Foo is an example field of StardogRole. Edit StardogRole_types.go to remove/update
-	Foo string `json:"foo,omitempty"`
+// StardogPermissionSpec defines a Stardog permission assigned to a Role
+type StardogPermissionSpec struct {
+    // Action describes the type of action a specific permission is assigned to
+    // Types of Actions: ALL, CREATE, DELETE, READ, WRITE, GRANT, REVOKE, EXECUTE
+	Action string `json:"action,omitempty"`
+	// ResourceType describes the type of resource a specific permission is assigned to
+	// Types of Resources: DB, USER, ROLE, ADMIN, METADATA, NAMEDGRAPH, VIRTUALGRAPH, ICVCONTRAINTS
+	ResourceType string `json:"resourceType,omitempty"`
+	// Resource describes the name of the specific Resource.
+	Resource string `json:"resource,omitempty"`
```

Comment:
Trying to combine the K8s RBAC and Stardog terminologies here:
```suggestion
    // Verb describes the action a specific permission is assigned to
    // Types of Actions: ALL, CREATE, DELETE, READ, WRITE, GRANT, REVOKE, EXECUTE
	Verbs string `json:"verbs,omitempty"`
	// Resources describes the list of resource the Verb can modify.
	// Types of Resources: DB, USER, ROLE, ADMIN, METADATA, NAMEDGRAPH, VIRTUALGRAPH, ICVCONTRAINTS
	Resources string `json:"resources,omitempty"`
	// Objects is a list of permission objects that get each targeted by the Verbs and Resources properties.
	Objects []string `json:"objects,omitempty"`
```
https://www.stardog.com/docs/#_resources

Reply:
Why would you mix up the K8s RBAC architecture with the one from Stardog? Here we are defining a CRD with content for Stardog that will be used by Kubernetes Operator... I don't think we need to add or mix structure/content between K8s and Stardog.

Reply:
The only thing mismatching is `action` vs `verb`, the rest fits both worlds. Since `Action` is very similar to `verb`, I'm fine with that.

Reply:
> All the current objects seem to be missing a reference to which Stardog instance/URL they are targeted. I suggest to create an additional StardogRoleBinding CRD that connects the User, Role and a new `StardogInstance` object.
> The `StardogInstance` CRD could hold properties like:
> 
> * `instanceURL`: URL of the stardog instance
> * `adminSecretRef`: Reference to stardog admin user credentials
> 
> Specifying the instance only in the StardogRoleBinding object eliminates the need to configure an instance URL in each User and Role. Having it in the StardogRoleBinding, this allows to re-use the same users and roles for multiple stardog instances, even in different namespaces.

You want to make these CRDs **cluster** scope not **Namespaced** scope. Do we have any security issues with that? (ex. having a user password the same to multiple environments) With your proposal we could define _User_ and _Role_ **cluster** scope and the _StardogRoleBinding_ **namespaced** scope. I see a lot of redundancy with url and admin credentials in _StardogRoleBinding_.

My solution is bind the CRDs to a namespace. Having a Stardog instance running in a namespace allows the Operator to get the **url** and **admin credentials** from that namespace (service and secret using the default namings from the helm chart) when the first user is defined, avoiding creating a bind or a new CRD.

Reply:
Where did get that with cluster-scope? :smile: I was nowhere mentioning that :smile: 

I was referring to the same functionality with Rolebinding as in RBAC:
```yaml
kind: StardogRoleBinding
metadata:
  name: dev-rolebinding
  namespace: stardog-dev
subjects:
- name: my-stardog-user-in-dev
rules:
- name: my-stardog-role-in-prod
  kind: StardogRole
  namespace: stardog-prod
stardogInstance:
  serverUrl: https://servicename
  adminCredentialRef: my-superadmin-secret
```
as you can see, the StardogRoleBinding and the instance are still Namespace scoped, but could potentially refer to roles that were defined in another namespace
And yes, it kinda makes sense to have a `StardogClusterRole` CRD, but this could be added later without breaking changes. It basically helps to reduce "spec duplication". The `StardogRolebinding` is basically the CRD that glues all the user, stardogs and roles together.

Reply:
Your idea to reuse Stardog Users and Roles doesn't imply that these CRDs be defined at cluster level?  (I think it's not clean to reuse a user for instance from a specific namespace to another) Rolebinding CRD has to be at namespace level as there can be (almost all cases) only 1 Stardog Instance per Namespace. You haven't commented on my solution yet... Do you want to have a Zoom Call to clarify our differences?